src/Indigo/Contracts/Holdings/Types.hs

```diff
@@ -0,0 +1,128 @@
+-- SPDX-FileCopyrightText: 2020 TBD
+--
+-- SPDX-License-Identifier: LicenseRef-Proprietary
+{-# OPTIONS_GHC -Wno-orphans #-}
+module Indigo.Contracts.Holdings.Types
+  ( Parameter (..)
+  , Storage
+  , StorageFields (..)
+  , TokenMeta (..)
+  , dummyMeta
+  , mkStorage
+  )
+where
+
+import Indigo
+
+import Fmt (Buildable(..), (+|), (|+))
+
+import qualified Lorentz.Contracts.ManagedLedger as ML
+import Lorentz.TypeAnns (HasTypeAnn)
+
+type instance ErrorArg "notInTransferAdminRightsMode" = ()
+
+instance CustomErrorHasDoc "notInTransferAdminRightsMode" where
+  customErrClass = ErrClassActionException
+  customErrDocMdCause =
+    "Cannot accept admin rights before transfer process has been initiated \
+    \by calling transferAdminRights entrypoint."
+
+type instance ErrorArg "senderIsNotNewAdmin" = ()
+
+instance CustomErrorHasDoc "senderIsNotNewAdmin" where
+  customErrClass = ErrClassBadArgument
+  customErrDocMdCause =
+    "Cannot accept admin rights because the sender address is different from \
+    \the address passed to the transferadminRights entrypoint previously."
+
+type instance ErrorArg "invalidSafelistAddr" = ()
+
+instance CustomErrorHasDoc "invalidSafelistAddr" where
+  customErrClass = ErrClassBadArgument
+  customErrDocMdCause =
+    "New safelist address parameter type doesn't match expected one."
+
+type instance ErrorArg "nonTransferable" = ()
+
+instance CustomErrorHasDoc "nonTransferable" where
+  customErrClass = ErrClassBadArgument
+  customErrDocMdCause =
+    "Transferable flag is false. Transfers are prohibited."
+
+data StorageFields = StorageFields
+  { tokenMeta :: TokenMeta
+  , mbSafelistAddress :: Maybe EpAddress
+  , owner :: Address
+  , admin :: Address
+  , mbNewAdmin :: Maybe Address
+  , paused :: Bool
+  , transferable :: Bool
+  , totalSupply :: Natural
```

Comment:
It seems this replaces `totalMinted` and `totalBurned`, perhaps the specification should be updated?
It also refers to these to fields in the `getTotalSupply` description

Reply:
The specification is more of a draft, which should be replaced with autodoc one day, but I'll update

Reply:
The specification should be treated as "requirements", IMO.